compiler/rustc_metadata/src/creader.rs

```diff
@@ -737,6 +834,25 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
         Ok(crate_num_map)
     }
 
+    fn resolve_target_modifiers(
+        &mut self,
+        crate_root: &CrateRoot,
+        metadata: &MetadataBlob,
+        krate: CrateNum,
+    ) -> Result<TargetModifiers, CrateError> {
+        debug!("resolving target modifiers of external crate");
+        if crate_root.is_proc_macro_crate() {
+            return Ok(TargetModifiers::new());
+        }
```

Comment:
How about special casing proc macros during encoding rather than decoding? That saves some space in the crate metadata and matches what we do for other kinds of proc macro metadata.

Reply:
I think, proc macro during encoding is processed by `empty_proc_macro!(self)`. Is it true?
I may remove special casing in decoding, it looks unnecessary, yes.
```
fn encode_target_modifiers(&mut self) -> LazyArray<TargetModifier> {
    empty_proc_macro!(self);
    let tcx = self.tcx;
    self.lazy_array(tcx.sess.opts.gather_target_modifiers())
}
```

I used `crate_deps` as an origin, `encode_crate_deps` uses `empty_proc_macro` and `resolve_crate_deps` performs fast path for `is_proc_macro_crate`.